gapidapk/deviceinfo.go

```diff
@@ -25,15 +28,117 @@ import (
 )
 
 const (
-	sendDevInfoAction  = "com.google.android.gapid.action.SEND_DEV_INFO"
-	sendDevInfoService = "com.google.android.gapid.DeviceInfoService"
-	sendDevInfoPort    = "gapid-devinfo"
+	sendDevInfoAction           = "com.google.android.gapid.action.SEND_DEV_INFO"
+	sendDevInfoService          = "com.google.android.gapid.DeviceInfoService"
+	sendDevInfoPort             = "gapid-devinfo"
+	startServiceAttempts        = 3
+	portListeningTimeoutSeconds = 5 * time.Second
 )
 
 func init() {
 	adb.RegisterDeviceInfoProvider(fetchDeviceInfo)
 }
 
+// Returns true if the device is listening to sendDevInfoPort, false if not.
+// Error if failed at getting the port info.
+func devInfoPortListening(ctx context.Context, d adb.Device) (bool, error) {
+	var stdout bytes.Buffer
+	if err := d.Shell("cat", "/proc/net/unix").Capture(&stdout, nil).Run(ctx); err != nil {
+		return false, log.Errf(ctx, err, "Getting unix abstract port info...")
+	}
+	if strings.Contains(stdout.String(), sendDevInfoPort) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// startDevInfoService tries to start the fresh run of the package and start
+// the service to send device info.
+func startDevInfoService(ctx context.Context, d adb.Device, apk *APK) error {
+	ctx = log.Enter(ctx, "startDevInfoService")
+	var se error
+	var le error
+	var listening bool
+
+	action := apk.ServiceActions.FindByName(sendDevInfoAction, sendDevInfoService)
+	if action == nil {
+		return log.Err(ctx, nil, "Service intent was not found")
+	}
+
+	// Try to start service.
+	for i := 0; i < startServiceAttempts; i++ {
+		log.I(ctx, "Attempt to start service: %s", sendDevInfoService)
+		if se = d.StartService(ctx, *action); se != nil {
+			continue
+		}
+
+		stop_checking_port := make(chan bool)
+		timer := time.AfterFunc(portListeningTimeoutSeconds, func() {
+			stop_checking_port <- true
+			close(stop_checking_port)
+		})
+		func() {
+			select {
+			case <-stop_checking_port:
+				return
+			default:
+				log.I(ctx, "Checking port: %s on device %s", sendDevInfoPort, d.Instance().GetSerial())
+				listening, le = devInfoPortListening(ctx, d)
+				if le != nil || listening {
+					return
+				}
+				time.Sleep(time.Second)
+			}
+		}()
+		timer.Stop()
+		if listening {
+			return nil
+		}
+	}
+	if se != nil {
+		return se
+	}
+	if le != nil {
+		return le
+	}
+	return log.Errf(ctx, nil, "Run out of attempts: %v", startServiceAttempts)
+}
+
+// Checks the existence of VkGraphicsSpyLayer in the debug.vulkan.layers system
+// property. If found, strip the layer, otherwise keep the property unchanged.
+// Returns a property value recover callback and error.
+func stripVkGraphicsSpyLayer(ctx context.Context, d adb.Device) (func(), error) {
```

Comment:
So if I understand this all correctly, this is all attempting to avoid a race (or failed cleanup) of the `debug.vulkan.layers` property, set my tracing, that when left set screws up the device info fetching?

It would appear at first glance that we've added a plaster to try and unset the layer here, but we're still going to have race issues if you try taking a trace at the same time as this logic is executed.

It feels to me that we probably need something more global-lock like. Perhaps we could use the device's properties for a lock on `debug.vulkan.layers`?

Reply:
Right. I'm just trying to make sure device info is gathered when 'debug.vulkan.layers' does not contains VkGraphicsSpy. And we still have a race condition.

One problem is that, when we test with Robot tester, a 'gapit trace' and 'gapit devices query' are done with different 'gapis' process. Even if we add a lock in the global level, we still have multiple copies of locks.

To totally avoid race for robot tester, it seems we will need a lock at robot level, generally that means, only one Vulkan operation (trace/replay/report) can be processed at a time, which is not good (two devices can not be used at the same time even if their workload are totally independent from each other).

I think probably we can use 'temporary files' with busy loop as 'locks' for this?

Reply:
I'm not sure if you saw or understood my suggestion for using a custom `setprop` variable on the device for the lock. This would mean the lock persists beyond gapit, gapis or robot - which may be slightly dangerous if one of these processes crash with the lock held. 
If we can use a property for a lock, perhaps it can have a periodically updated timestamp so that times greater than some threshold are ignored. 
Of course you could also use a temporary file, but properties have the nice property (pun not intended) of being wiped between reboots.

Reply:
Just uploaded 3a5af05 as a preparation of using 'property' for 'lock'. (we don't want 'trace' to block 'deviceinfo query', as 'trace' takes a long time, but we can have 'deviceinfo query' to block 'trace').

Trace is responsible to set and unset 'debug.vulkan.layers'
DeviceInfo query is responsible to set and unset 'devinfo-querying'

Trace can not happen if a 'devinfo-querying' property is set.
DeviceInfo query can not happen if 'debug.vulkan.layers' is set.

But I'm thinking using a system property probably won't resolve the whole problem.
For robot tester, there will never be two trace happening at the same time, but multiple deviceinfo query may happen at the same time (this is because `gapit devices` and its underlying service traverse all the connected devices, and robot tester will spawn individual gapis for each test).

So, A pending 'trace' (which is waiting for devinfo query X may still get a race condition with a 'deviceinfo query' Y, which is initiated after X.

Is there a good solution for this problem, I thought I need an 'atomic getAndSetSystemProperty' thing...